ws/test_ws_ping.py

```diff
@@ -0,0 +1,334 @@
+    default Upgrade;
+    ''      close;
+}
+
+upstream websocket {
+    ip_hash;
+    server 127.0.0.1:8099;
+}
+
+upstream wss_websockets {
+    ip_hash;
+    server 127.0.0.1:8099;
+}
+
+server {
+    listen 8000;
+    location / {
+        proxy_pass http://websocket;
+        proxy_http_version 1.1;
+        proxy_set_header Upgrade $$http_upgrade;
+        proxy_set_header Connection $$connection_upgrade;
+        proxy_set_header Host $$host;
+    }
+}
+
+server {
+    listen 8001 ssl;
+    ssl_protocols TLSv1.2;
+    ssl_certificate /tmp/cert.pem;
+    ssl_certificate_key /tmp/key.pem;
+    location / {
+        proxy_pass http://wss_websockets;
+        proxy_http_version 1.1;
+        proxy_set_header Upgrade $$http_upgrade;
+        proxy_set_header Connection $$connection_upgrade;
+        proxy_set_header Host $$host;
+    }
+}
+"""
+
+def cert_gen(
+        emailAddress="emailAddress",
+        commonName=hostname,
+        countryName="NT",
+        localityName="localityName",
+        stateOrProvinceName="stateOrProvinceName",
+        organizationName="organizationName",
+        organizationUnitName="organizationUnitName",
+        serialNumber=0,
+        validityStartInSeconds=0,
+        validityEndInSeconds=10*365*24*60*60,
+        KEY_FILE="/tmp/key.pem",
+        CERT_FILE="/tmp/cert.pem"):
+    haproxy_pem = "/tmp/out.pem"
+    out_files = []
+    k = crypto.PKey()
+    k.generate_key(crypto.TYPE_RSA, 4096)
+    cert = crypto.X509()
+    cert.get_subject().C = countryName
+    cert.get_subject().ST = stateOrProvinceName
+    cert.get_subject().L = localityName
+    cert.get_subject().O = organizationName
+    cert.get_subject().OU = organizationUnitName
+    cert.get_subject().CN = commonName
+    cert.get_subject().emailAddress = emailAddress
+    cert.set_serial_number(serialNumber)
+    cert.gmtime_adj_notBefore(0)
+    cert.gmtime_adj_notAfter(validityEndInSeconds)
+    cert.set_issuer(cert.get_subject())
+    cert.set_pubkey(k)
+    cert.sign(k, 'sha512')
+    with open(CERT_FILE, "wt") as f:
+        f.write(crypto.dump_certificate(
+            crypto.FILETYPE_PEM, cert).decode("utf-8"))
+        out_files.append(CERT_FILE)
+    with open(KEY_FILE, "wt") as f:
+        f.write(crypto.dump_privatekey(crypto.FILETYPE_PEM, k).decode("utf-8"))
+        out_files.append(KEY_FILE)
+    filenames = [CERT_FILE, KEY_FILE]
+    with open('/tmp/out.pem', 'w') as outfile:
+        for fname in filenames:
+            with open(fname) as infile:
+                for line in infile:
+                    outfile.write(line)
+    out_files.append(haproxy_pem)
+    return out_files
+
+
+def remove_certs(cert_files_):
+    for cert in cert_files_:
+        os.remove(cert)
+
+
+class Ws_ping(tester.TempestaTest):
+
+    backends = []        
+
+    clients = []
+
+    tempesta = {
+        'config': TEMPESTA_CONFIG % "",
+    }
+
+    # Client
+
+    def run_test(self, port, n):
+        time.sleep(2.0)
```

Comment:
2 seconds sleeping is too much. We already have issues with the current timeouts https://github.com/tempesta-tech/tempesta-test/issues/120 , in different aspect, but timeouts hurt tests performance badly.

Please use events and conditional waits wherever possible

Reply:
fixed through:
self.get_tempesta().is_running()